plugin-gen3/auth_impl.go

```diff
@@ -3,97 +3,251 @@ package main
-	shared.Logger.Info("Configuration", "OidcTokenUrl", pluginConfig.OidcTokenUrl)
-	shared.Logger.Info("Configuration", "OidcClientId", pluginConfig.OidcClientId)
+	scopes := []string{"openid"}
+	expected := &authutils.Expected{Scopes: scopes}
 
-	// exchange the OIDC client ID and secret for an access token
+	err = expected.Validate(claims)
+	if err != nil {
+		return "", fmt.Errorf("error decoding token: %w", err)
+	}
+	userIdInterface, exists := (*claims)["sub"]
+	if !exists {
+		return "", missingRequiredField("sub")
+	}
+	userId, casted := userIdInterface.(string)
+	if !casted {
+		return "", fieldTypeError("sub")
+	}
+
+	return userId, nil
+}
+
+func (a Authorize) PluginAction(params map[string]string, headers map[string]*proto.StringList, configuration *config.Config, task *tes.Task, taskType proto.Type) (*proto.JobResponse, error) {
+	// only proceed for task creation events. The worker config does not need to be updated for
+	// other types of events
+	if taskType == proto.Type_GET || taskType == proto.Type_CANCEL {
+		return &proto.JobResponse{Code: http.StatusOK, Config: configuration, Task: task}, nil
+	}
+	if taskType != proto.Type_CREATE {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: fmt.Sprintf("unsupported task type: %v", taskType),
+			},
+			fmt.Errorf("unsupported task type: %v", taskType)
+	}
+
+	// get the plugin configuration
+	// The OIDC client should be created in Gen3 with:
+	// `fence-create client-create --client CLIENT_NAME --grant-types client_credentials`
+	S3Url, ok := params["S3Url"]
+	if !ok || S3Url == "" {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: "S3Url is required in params",
+			},
+			fmt.Errorf("S3Url is required in params")
+	}
+	OidcClientId, ok := params["OidcClientId"]
+	if !ok || OidcClientId == "" {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: "OidcClientId is required in params",
+			},
+			fmt.Errorf("OidcClientId is required in params")
+	}
+	OidcClientSecret, ok := params["OidcClientSecret"]
+	if !ok || OidcClientSecret == "" {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: "OidcClientSecret is required in params",
+			},
+			fmt.Errorf("OidcClientSecret is required in params")
+	}
+	shared.Logger.Info("Configuration", "S3Url", S3Url, "OidcClientId", OidcClientId)
+
+	// get the user's access token from the headers
+	authHeaders, ok := headers["authorization"]
+	if !ok || authHeaders == nil || len(authHeaders.Values) == 0 {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: "Authorization header is required",
+			},
+			fmt.Errorf("Authorization header is required")
+	}
+	authHeader := authHeaders.Values[0]
+	if authHeader == "" {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: "Authorization header is required",
+			},
+			fmt.Errorf("Authorization header is required")
```

Comment:
These lines for BadRequest are being repeated multiple times and I feel like they're adding a lot of fluff to the code. How about we write a function that takes in a message and returns a JobResponse object. Something like this (untested)

```
func badRequestJobResponse(msg string) (*proto.JobResponse, error) {
	return &proto.JobResponse{
			Code:    http.StatusBadRequest,
			Message: msg,
		},
		fmt.Errorf(msg)
}
```

and simply call this function everywhere where we are returning a BadRequest status 


```suggestion
		return badRequestJobResponse("Authorization header is required")
```

Reply:
Or maybe a more generic helper function like
```
// Helper function to create a JobResponse with a customizable error code and error message.
// If err is nil, a new error is created from msg.
func jobResponseError(code int, msg string, err error) (*proto.JobResponse, error) {
    if err == nil {
        err = fmt.Errorf(msg)
    } else {
        err = fmt.Errorf("%s: %w", msg, err)
    }
    return &proto.JobResponse{
            Code:    code,
            Message: msg,
        },
        err
}


// Example usage:
if err != nil {
    msg := fmt.Sprintf("error making HTTP request to '%s'", url)
    return jobResponseError(http.StatusInternalServerError, msg, err)
}